bundles/org.openhab.binding.energidataservice/src/main/java/org/openhab/binding/energidataservice/internal/action/EnergiDataServiceActions.java

```diff
@@ -0,0 +1,381 @@
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.energidataservice.internal.PriceCalculator;
+import org.openhab.binding.energidataservice.internal.exception.MissingPriceException;
+import org.openhab.binding.energidataservice.internal.handler.EnergiDataServiceHandler;
+import org.openhab.core.automation.annotation.ActionInput;
+import org.openhab.core.automation.annotation.ActionOutput;
+import org.openhab.core.automation.annotation.RuleAction;
+import org.openhab.core.library.types.QuantityType;
+import org.openhab.core.library.unit.Units;
+import org.openhab.core.thing.binding.ThingActions;
+import org.openhab.core.thing.binding.ThingActionsScope;
+import org.openhab.core.thing.binding.ThingHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link EnergiDataServiceActions} provides actions for getting energy data into a rule context.
+ *
+ * @author Jacob Laursen - Initial contribution
+ */
+@ThingActionsScope(name = "energidataservice")
+@NonNullByDefault
+public class EnergiDataServiceActions implements ThingActions {
+
+    private final Logger logger = LoggerFactory.getLogger(EnergiDataServiceActions.class);
+
+    private @Nullable EnergiDataServiceHandler handler;
+
+    private enum PriceElement {
+        SPOT_PRICE("spotprice"),
+        NET_TARIFF("nettariff"),
+        SYSTEM_TARIFF("systemtariff"),
+        ELECTRICITY_TAX("electricitytax"),
+        TRANSMISSION_NET_TARIFF("transmissionnettariff");
+
+        private static final Map<String, PriceElement> NAME_MAP = Stream.of(values())
+                .collect(Collectors.toMap(PriceElement::toString, Function.identity()));
+
+        private String name;
+
+        private PriceElement(String name) {
+            this.name = name;
+        }
+
+        @Override
+        public String toString() {
+            return name;
+        }
+
+        public static PriceElement fromString(final String name) {
+            PriceElement myEnum = NAME_MAP.get(name.toLowerCase());
+            if (null == myEnum) {
+                throw new IllegalArgumentException(String.format("'%s' has no corresponding value. Accepted values: %s",
+                        name, Arrays.asList(values())));
+            }
+            return myEnum;
+        }
+    }
+
+    @RuleAction(label = "@text/action.get-prices.label", description = "@text/action.get-prices.description")
+    public @ActionOutput(name = "prices", type = "java.util.Map<java.time.Instant, java.math.BigDecimal>") Map<Instant, BigDecimal> getPrices() {
+        return getPrices(Arrays.stream(PriceElement.values()).collect(Collectors.toSet()));
+    }
+
+    @RuleAction(label = "@text/action.get-prices.label", description = "@text/action.get-prices.description")
+    public @ActionOutput(name = "prices", type = "java.util.Map<java.time.Instant, java.math.BigDecimal>") Map<Instant, BigDecimal> getPrices(
```

Comment:
@florian-h05 - sorry to bother you again. I'm progressing with baby steps migrating to JS. Calling this method:

```javascript
edsActions = actions.get("energidataservice", "energidataservice:service:energidataservice");
var priceMap = edsActions.getPrices();
console.log("getPrices" + priceMap.toString());
```

will print:

```
2023-05-17 22:37:58.829 [INFO ] [enhab.automation.script.file.test.js] - getPrices{2023-05-17T14:00:00Z=0.595739002, 2023-05-17T10:00:00Z=0.473758999, 2023-05-18T10:00:00Z=0.915008991, 2023-05-18T06:00:00Z=1.119208973, 2023-05-18T02:00:00Z=1.149668995, 2023-05-17T22:00:00Z=1.195908985, 2023-05-17T18:00:00Z=1.298829029, 2023-05-18T18:00:00Z=1.384179005, 2023-05-18T14:00:00Z=0.979718983, 2023-05-17T15:00:00Z=0.886108998, 2023-05-17T11:00:00Z=0.447169, 2023-05-18T07:00:00Z=1.045108998, 2023-05-18T03:00:00Z=1.179598988, 2023-05-17T23:00:00Z=1.15353899, 2023-05-17T19:00:00Z=1.392889027, 2023-05-18T19:00:00Z=1.319458973, 2023-05-18T15:00:00Z=1.076759022, 2023-05-18T11:00:00Z=0.770899006, 2023-05-17T16:00:00Z=0.931019001, 2023-05-17T12:00:00Z=0.447399, 2023-05-17T08:00:00Z=0.80828899, 2023-05-18T08:00:00Z=0.983449024, 2023-05-18T04:00:00Z=1.194648976, 2023-05-18T00:00:00Z=1.111389027, 2023-05-17T20:00:00Z=1.253629017, 2023-05-18T20:00:00Z=1.263088978, 2023-05-18T16:00:00Z=1.214458973, 2023-05-18T12:00:00Z=0.665819004, 2023-05-17T13:00:00Z=0.461839, 2023-05-17T09:00:00Z=0.717728993, 2023-05-18T09:00:00Z=0.931089009, 2023-05-18T05:00:00Z=1.153018971, 2023-05-18T01:00:00Z=1.109009022, 2023-05-17T21:00:00Z=1.195018971, 2023-05-17T17:00:00Z=0.971458973, 2023-05-18T21:00:00Z=1.18228899, 2023-05-18T17:00:00Z=1.307179005, 2023-05-18T13:00:00Z=0.767628987}
```

But the type seems to be `java.util.HashMap` - which is correct, even if exposed as interface `java.util.Map`. I guess I was somehow expecting a JS `Map`.

Clearly I received the correct data, but I'm not sure how to process it on the JS side. For example, I cannot perform a lookup by key having a js-joda `Instant`. Lookup by string key doesn't work either because of type native Java type.

Do you have any recommendations here?

Reply:
> But the type seems to be java.util.HashMap - which is correct, even if exposed as interface java.util.Map. I guess I was somehow expecting a JS Map.

That’s the direction that I haven’t a way to do type mappings yet …
openhab-js is internally doing manual Java to JS type mappings, they are available on the `utils` namespace. You could use `utils.javaMapToJsMap` or `utils.javaMapToJsObj`, see https://openhab.github.io/openhab-js/utils.html#.javaMapToJsMap.

For the java.time.Instant - JS-Joda Instant problem, you can either copy the JS type and have the Instants converted to JS-Joda Instant or stringified.

However I’d have to try that myself to give real advice, but this could get a bit complicated.

If you want a Map with the Prices for each hour, I’d rather recommend to create that on the JS side. Use a JS-Joda Instant, get the price for it, put the instant and the price into a js map and so on …

Reply:
Thanks again. I started adding some example JavaScript code in this commit: d512666710debfcb6072d68c3bd8c511f908df1f

`utils.javaMapToJsMap` did the trick, except that keys were still Java `Instant`, so I converted the `Map` to have string keys. I tried also with JS-Joda `Instant` but couldn't make that work. I believe they are immutable, but maybe the `Map` itself doesn't support using such objects as keys. At least now that example works, and it can be improved as I learn more about JavaScript. 🙂

Reply:
The code looks good and simple.


> I tried also with JS-Joda Instant but couldn't make that work. 

What exactly didn't work?

> the Map itself doesn't support using such objects as keys.

It should accept any value as key, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map.

Reply:
You are right, I tried again, this time with better luck:

```javascript
utils.javaMapToJsMap(edsActions.getPrices()).forEach((value, key) => {
    priceMap.set(time.Instant.parse(key.toString()), value);
});
var hourStart = time.Instant.now().truncatedTo(time.ChronoUnit.HOURS);
console.log("Current total price excl. VAT: " + priceMap.get(hourStart));
```

A bit clumsy (and probably slow) to use `Instant.parse`, don't know if there is a more straight path from `java.time.Instant` to js-joda `Instant`.

Reply:
> A bit clumsy (and probably slow) to use Instant.parse, don't know if there is a more straight path from java.time.Instant to js-joda Instant.

I think there is a better way, I’ll check that myself and add this to the `utils` so the user doesn’t have to care about the implementation.

Reply:
I found a much faster way.
Using `Instant.parse` takes 150-200 ms for 50 iterations.
With my approach, 50 iteration only take 7-10 ms.

My benchmark script:

```javascript
utils.javaInstantToJsInstant = function (java) {
  return time.Instant.ofEpochMilli(java.toEpochMilli());
}

var Instant = Java.type('java.time.Instant');
var ins = Instant.now();

console.log('Benchmarking Instant.parse conversion ...')
var start = time.Instant.now().toEpochMilli();

for (var i = 0; i < 50; i++) {
  time.Instant.parse(ins.toString())
}

var end = time.Instant.now().toEpochMilli();
console.log('Benchmark finshed, took ' + (end - start) + ' ms')

console.log('Benchmarking advanced Instant conversion ...')
var start = time.Instant.now().toEpochMilli();

for (var i = 0; i < 50; i++) {
  utils.javaInstantToJsInstant(ins)
}

var end = time.Instant.now().toEpochMilli();
console.log('Benchmark finshed, took ' + (end - start) + ' ms')
```

I will open a PR for the JS Scripting library openhab-js to add this utility function as well as one for ZonedDateTime.

Reply:
@jlaur PR is open: https://github.com/openhab/openhab-js/pull/267.

Reply:
@florian-h05 - your optimized utility method works well.

However, I have to back-paddle on this: https://github.com/openhab/openhab-addons/pull/14376#discussion_r1199656891. It does not work. The reason is quite simple: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map and https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality. So the key only matches if it's the same object (instance), i.e. `===` is used, not `equals`.

So this is just to conclude that this example should either stay as is with string representation as keys, or alternatively use epoch milliseconds.

Reply:
> your optimized utility method works well.

Great to hear 👍 

> However, I have to back-paddle on this: https://github.com/openhab/openhab-addons/pull/14376#discussion_r1199656891. It does not work.

Makes sense, unfortunately for objects the lookup is also using `===`, so you need to stick with primitive types like string or number.